### PR TITLE
swap_ranges, bitexch, and warning fixes

### PR DESCRIPTION
**Pull request by bkille**
`write_word` and the end-iterator guarded overloaded `get_word` still need tests. I can remove them from the PR if desired.

### CODE REVIEW

File: include/bit_algorithm_details.hpp [hunk 1]
```diff
@@ -92,6 +96,138 @@ T get_word(bit_iterator<InputIt> first, T len=binary_digits<T>::value)
...
+T get_word(bit_iterator<InputIt> first, bit_iterator<InputIt> last,
+        T& bits_read, T len=binary_digits<T>::value
+        )
+{
+    using native_word_type = typename bit_iterator<InputIt>::word_type;
+    constexpr T native_digits = binary_digits<native_word_type>::value; 
+    constexpr T ret_digits = binary_digits<T>::value; 
+    assert(ret_digits >= len);
+    bits_read = native_digits - first.position();
+    T ret_word = *first.base() >> first.position();
```
Comment (gress2) [inline include/bit_algorithm_details.hpp]: i'm scared of this now. if `*first.base()` is a short there might be weird behavior?

File: include/bit_algorithm_details.hpp [hunk 2]
```diff
@@ -92,6 +96,138 @@ T get_word(bit_iterator<InputIt> first, T len=binary_digits<T>::value)
...
+
+    // TODO vincent mentioned that we should aim for only 1 return function
```
Comment (gress2) [inline include/bit_algorithm_details.hpp]: Do you remember why he said this? For the sake of software engineering or performance?

File: include/bit_algorithm_details.hpp [hunk 3]
```diff
@@ -92,6 +96,138 @@ T get_word(bit_iterator<InputIt> first, T len=binary_digits<T>::value)
...
+template <class T, class InputIt>
+T get_word(bit_iterator<InputIt> first, bit_iterator<InputIt> last,
+        T& bits_read, T len=binary_digits<T>::value
+        )
+{
+    using native_word_type = typename bit_iterator<InputIt>::word_type;
+    constexpr T native_digits = binary_digits<native_word_type>::value; 
+    constexpr T ret_digits = binary_digits<T>::value; 
+    assert(ret_digits >= len);
+    bits_read = native_digits - first.position();
```
Comment (gress2) [inline include/bit_algorithm_details.hpp]: really small thing but can we move the `bits_read` assignment to be after the actual read

File: include/bit_algorithm_details.hpp (hunk 1 above)
Comment (bkille) [reply include/bit_algorithm_details.hpp]: So I think here it is immediately casted to to `T`, so there is no issue. I think its when you use the results in intermediate operations where issues arise.

File: include/bit_algorithm_details.hpp (hunk 2 above)
Comment (bkille) [reply include/bit_algorithm_details.hpp]: I believe it's strictly S/E. I've also seen it in a Best Practices book. However its obviously not worth it if it causes the function to be even less readable.

File: include/rotate.hpp [hunk 4]
```diff
@@ -2,23 +2,56 @@
...
+bit_iterator<ForwardIt> rotate(
+   bit_iterator<ForwardIt> first, 
+   bit_iterator<ForwardIt> n_first,
+   bit_iterator<ForwardIt> last
+) {
+    return __rotate(
+            first, 
+            n_first, 
+            last, 
+            typename std::iterator_traits<ForwardIt>::iterator_category()
```
Comment (gress2) [inline include/rotate.hpp]: what happens when this isn't a RAI? still to do i guess?

Comment (bkille) [reply include/rotate.hpp]: Yes, I'll put the other stubs in as todo

**Pull request closed without merging**